delfin/leader_election/tooz/leader_elector.py

```diff
@@ -0,0 +1,120 @@
+
+import threading
+
+from oslo_log import log
+from oslo_utils import timeutils
+
+from delfin.coordination import LeaderElectionCoordinator
+from delfin.leader_election.interface import LeaderElector
+
+LOG = log.getLogger(__name__)
+
+
+class Elector(LeaderElector):
+
+    def __init__(self, callbacks, leader_election_key):
+        key = leader_election_key.encode('ascii')
+        super(Elector, self).__init__(callbacks, key)
+
+        self._coordinator = None
+        self.leader = False
+        self._stop = threading.Event()
+        self._runner = None
+        self._beats = 0
+
+    def run(self):
+        if self._coordinator:
+            return
+
+        self._stop.clear()
+        self._beats = 0
+
+        self._coordinator = LeaderElectionCoordinator()
+        # start leader coordinator
+        self._coordinator.start()
+
+        self._coordinator.ensure_group(self.election_key)
+        # join coordinator
+        self._coordinator.join_group()
+
+        # register callback for elected leader
+        self._coordinator. \
+            register_on_start_leading_callback(self.
+                                               callbacks.on_started_leading)
+
+        # register internal callback to notify being a leader
+        self._coordinator. \
+            register_on_start_leading_callback(self.set_leader_callback)
+
+        while not self._stop.is_set():
+            with timeutils.StopWatch() as w:
+                LOG.info("sending heartbeats for leader election")
+                wait_until_next_beat = self._coordinator.start_heartbeat()
+
+            ran_for = w.elapsed()
+            has_to_sleep_for = wait_until_next_beat - ran_for
+            if has_to_sleep_for < 0:
+                LOG.warning(
+                    "Heartbeating took too long to execute (it ran for"
+                    " %0.2f seconds which is %0.2f seconds longer than"
+                    " the next heartbeat idle time). This may cause"
+                    " timeouts (in locks, leadership, ...) to"
+                    " happen (which will not end well).", ran_for,
+                    ran_for - wait_until_next_beat)
+            self._beats += 1
+
+            # check if coordinator is still a leader
+            if self.leader and not self._coordinator.is_still_leader():
+                self.on_stopped_leading()
+                self.leader = False
+                return
+            self._coordinator.start_leader_watch()
+
+            if self.leader:
+                # adjust time for leader
+                has_to_sleep_for = has_to_sleep_for / 2
+
+            LOG.info('resting after leader watch as leader=%(leader)s '
+                     'for heartbeat timeout of %(timeout)s sec',
+                     {'timeout': has_to_sleep_for, 'leader': self.leader})
+
+            self._stop.wait(has_to_sleep_for)
+
+    def set_leader_callback(self, *args, **kwargs):
+        self.leader = True
+
+    def cleanup(self):
```

Comment:
can we give a docstring here to describe what the function does?

Reply:
cleanup is abstract method of ```LeaderElector``` interface. The docstring is already specified there.